### ISSUE TITLE
Using prebuilt images should work without needing a `Dockerfile`.

### ISSUE THREAD

**Issue by alexdima** [microsoft/vscode-dev-containers#220]
Testing https://github.com/microsoft/vscode-remote-release/issues/2209

* Run `Remote-Containers: Add Development Container Configuration Files...`
* Pick the Node.js 12 & TypeScript

Currently, there is a `Dockerfile` being generated and a `devcontainer.json`. They both contain a lot of stuff about the current user id and a lot of comments and explanations and this scares me away and makes me feel that this is not for me.

I would like to suggest that we make it possible that we don't have a `Dockerfile` and maybe point to the image directly from `devcontainer.json`. IMHO, it is so cool that we now have prebuilt images, and I think we should support a dead simple variant where I can say: "hey VS Code, I want to try out Node.js 12 & TypeScript" and that results in a single file `devcontainer.json` and that file has only a few lines of text in it. Maybe the implementation still uses a `Dockerfile` under the hood, but this `Dockerfile` is not presented to the user and is not present in the repository until an advanced customization warrants it.

Currently, I feel so overloaded with information and I do not know how to act on the information and why it needs to be in my project, because I just wanted a simple, standard image, without any advanced user id mapping.:

<details><summary>Wall of text</summary>

   ![image](https://user-images.githubusercontent.com/5047891/73276144-80f55000-41e8-11ea-9526-47d500330ff8.png)

   ![image](https://user-images.githubusercontent.com/5047891/73276382-e47f7d80-41e8-11ea-8996-62d44350344e.png)

</details>

I personally think there are all kinds of users:
- experts in docker. They feel comfortable editing `Dockerfile`s and are not shy to go deep.
- beginners in docker that open a folder where an expert has written the `Dockerfile`.
- others (vast majority of people) who want to try out what we have done, but might be intimidated by the complexity of docker and `Dockerfile` and feel paralysed by the sheer amount of possibilities. 

I would like that we target also the last group of folks who would like to try dev containers in a "standard way"... For example I would like to learn Go and solve some programming problems in Go, but now I first need to learn all this stuff about Dockerfile and I never get to learn Go because I am stuck learning about GUID and UID stuff...

I think we are very very close to having a great experience...

**Comment by Chuxel** [comment]
@alexdima The only reason the UID/GID section is present is because we have not implemented dynamic UID/GID (updateRemoteUserUID) for Linux when using Docker Compose.  Once that is in place, we'd remove that entirely from these files.  It's essentially a short term workaround.  This is unfortunately a critical thing for Linux since if you don't do this, all files you edit will be done so as root.

The ultimate goal here would be to have a single Dockerfile that is empty, with simply some commented out things to get you started on installing additional software. The intent of these is to be used with actual projects rather than as samples, and in the majority case, people are going to want to install other things over time. If you are new to Linux, you won't know things like apt-get, so there'd just be enough there to let you know how to do it, but all commented out.

The comments in devcontainer.json have been slowly added over time based on the questions most people have asked when they start using the product.  There again, once we get the UID/GID support in Docker Compose, some of it can clearly be paired down.

The goal is to really answer the question, "what next?" Originally none of these things were there and we got quite a few additional questions on basic things everyone wants to know like exposing ports.

**Comment by Chuxel** [comment]
@alexdima I should also ask, given the context above, are there things in that you think aren't needed?  One option, for example, would be to default to a non-root user and simply drop the comment assuming that the property "remoteUser" is self explanatory enough.

**Comment by Chuxel** [comment]
@alexdima I queued up a PR to try to reduce the amount of comments while still retaining the core info that has seemed to be confusing to people as they first start using devcontainer.json.

See if you think it does a better job of balancing.  For this particular image, here's what you would now get.

devcontainer.json
```json
{
	"name": "Node.js 12 & TypeScript",
	"dockerFile": "Dockerfile",

	// Set *default* container specific settings.json values on container create.
	"settings": { 
		"terminal.integrated.shell.linux": "/bin/bash"
	},

	// Add the IDs of extensions you want installed when the container is created.
	"extensions": [
		"dbaeumer.vscode-eslint",
		"ms-vscode.vscode-typescript-tslint-plugin"
	]

	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	// "forwardPorts": [],

	// Use 'postCreateCommand' to run commands after the container is created.
	// "postCreateCommand": "yarn install",

	// Uncomment to connect as a non-root user. See https://aka.ms/vscode-remote/containers/non-root.
	// "remoteUser": "node"
}
```

Dockerfile

```Dockerfile
#-------------------------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See https://go.microsoft.com/fwlink/?linkid=2090316 for license information.
#-------------------------------------------------------------------------------------------------------------

FROM mcr.microsoft.com/vscode/devcontainers/typescript-node:12

# ** [Optional] Uncomment this section to install additional packages. **
#
# ENV DEBIAN_FRONTEND=noninteractive
# RUN apt-get update \
#    && apt-get -y install --no-install-recommends <your-package-list-here> \
#    #
#    # Clean up
#    && apt-get autoremove -y \
#    && apt-get clean -y \
#    && rm -rf /var/lib/apt/lists/*
# ENV DEBIAN_FRONTEND=dialog
```

**Comment by alexdima** [comment]
Thank you @Chuxel! Did I mention how cool the prebuilt images are? I am super excited that I can now rebuild containers in <5s !! :heart: 

> I have opened https://github.com/microsoft/vscode-remote-release/issues/2227 so we try to further reduce the rebuild time. FYI @egamma it would be awesome to get that on the February plan. Once we do that, the times might be <2s and then we can look for what else is needed to reduce the time further. IMHO there will be a magic number like <1s and once we are there, might make containerised development more mainstream. i.e. if my project loads up in <1s, then I'm willing to switch to use docker containers for day-to-day development.

Back on topic, with your proposed changes, these files look now a whole lot better IMHO! But there are still some nitpicks to get this to the minimal size and perhaps some other small things that might keep people from jumping on this.

Please forgive me if I am a noob and my suggestions are bad :).

1. The MS copyright statement in the `Dockerfile`. IMHO open source projects might not feel comfortable adding a file with a foreign copyright to their projects. At this point, there is one line in that file that is not a comment, so it might be reasonable to remove the copyright statement. I think we have some clauses that templates/examples can be open sourced without a copyright statement, since we would never want to actually enforce the copyright for templates/examples anyways. So I would suggest to delete this:
```
#-------------------------------------------------------------------------------------------------------------
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See https://go.microsoft.com/fwlink/?linkid=2090316 for license information.
#-------------------------------------------------------------------------------------------------------------
```

2. Can this be simplified? I think this is the most complete usage of apt-get I have seen in quite a while. Is it really necessary to trim the image at the end since this is not a production image? Also, in this case, IMHO it would be far more reasonable to show how to install gcc / node-gyp since this is a container for node, it is likely that someone wishes to do a native node module. Another good sample might be how to install `yarn`:
```
# ENV DEBIAN_FRONTEND=noninteractive
# RUN apt-get update \
#    && apt-get -y install --no-install-recommends <your-package-list-here> \
#    #
#    # Clean up
#    && apt-get autoremove -y \
#    && apt-get clean -y \
#    && rm -rf /var/lib/apt/lists/*
# ENV DEBIAN_FRONTEND=dialog
```
==>
(please check if this is really working, I just typed it on GitHub's textarea and the base image might already have python so this might be redundant :D)
```
# Uncomment to install dependencies for developing native node modules:
RUN apt-get -y install python make gcc g++
# Uncomment to install yarn
RUN apt-get ???
```
Now that I just typed the above, I think that these might make rebuilding the container a lot slower, right? We might want to offer prebuilt variants that contain yarn and native node module tools.

Actually, the more I think about it, a separate template might make more sense because the native node module variant might benefit from the C++ extension.

So I guess I am suggesting two things:
- to change the comment to show how to install yarn (or perhaps simply have yarn in the base image if the size cost is not too great).
- to create another image which is perfectly tailored for native node module development

3. Terminal settings
```
	// Set *default* container specific settings.json values on container create.
	"settings": { 
		"terminal.integrated.shell.linux": "/bin/bash"
	},
```
Why is the above needed? Why is our built-in terminal not picking `/bin/bash` by default? IMHO we should fix things inside VS Code such that the terminal picks the best available shell, and not the most common one. Perhaps an issue already exists that we could prioritize?

4. JSON + Comments
```
	// Use 'forwardPorts' to make a list of ports inside the container available locally.
	// "forwardPorts": [],

	// Use 'postCreateCommand' to run commands after the container is created.
	// "postCreateCommand": "yarn install",

	// Uncomment to connect as a non-root user. See https://aka.ms/vscode-remote/containers/non-root.
	// "remoteUser": "node"
```
AFAICT, we have never done this for JSON. We have amazing good suggestions to help you find whatever you want to add.

---

A note and explanation to my thinking and what is my angle:

IMHO I would find it awesome if this entire thing could be reduced to a single `devcontainer.json` file (even without a `Dockerfile`):

a single `devcontainer.json`:
```
{
  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:12"
}
```
I am not 100% certain about the `"extensions"` field. IMHO it would be desirable to drop that as well and perhaps they could be read from the workspace recommendations if they are not already.

So, to make a project be containerized, I would need to add this `devcontainer.json`:
```
{
  "image": "mcr.microsoft.com/vscode/devcontainers/typescript-node:12"
}
```

And that's it! I can imagine making a PR to some popular OS projects and ask them to become containerized with a one liner... And, if the container opens in <1s, they might just take out PR :)... IMHO, for node modules, from what I have seen, I think this might cover 80% of projects. Then there are all kind of complications, like projects that need additional `apt` dependencies, etc. etc. IMHO only those projects should be the ones that need to write a `Dockerfile`

To sum up, I think we should aim for "one-liners" containers. The benefits are multiple:
1. small footprint and a lot easier to add to open source projects
2. it is easy for someone to switch between base images, using ctrl+space they can switch between a base image with node 12 and one with node 10, or one with g++
3. amazing fast speed if a custom Dockerfile is avoided

In any way, thanks for reading so far down! :heart:

**Comment by Chuxel** [comment]
> The MS copyright statement in the Dockerfile. IMHO open source projects might not feel comfortable adding a file with a foreign copyright to their projects. At this point, there is one line in that file that is not a comment, so it might be reasonable to remove the copyright statement. I think we have some clauses that templates/examples can be open sourced without a copyright statement, since we would never want to actually enforce the copyright for templates/examples anyways. So I would suggest to delete this

I can log an enhancement to remove it when it is applied. Since it's OSS, it needs to be there in the repo.

> Can this be simplified? I think this is the most complete usage of apt-get I have seen in quite a while. Is it really necessary to trim the image at the end since this is not a production image? Also, in this case, IMHO it would be far more reasonable to show how to install gcc / node-gyp since this is a container for node, it is likely that someone wishes to do a native node module. Another good sample might be how to install yarn

What this is illustrating is best practice. This handles three problems when dealing with Docker and `apt-get`.
1. `apt-get` won't work in a container unless you do `apt-get update` first
2. `--no-install-recommends` keeps the image size down by not installing packages you did not explicitly ask for and their dependencies.
2. To keep the image size down, you need to clean out the cache when you are done.
3. You will see errors when installing without setting `ENV DEBIAN_FRONTEND=noninteractive`

All of this is standard in Dockerfiles. These files need to be as much about helping people understand what to do as getting going quickly.

Yarn, python, zsh, tslint, eslint, tsc and more is already in the image for Node. This is for libraries that the developer needs that are specific to their scenario.  For example, they might want to use the fish shell.  This image is already ~900MB (much of which is the Node image itself), so its pretty packed with content.

> I am not 100% certain about the "extensions" field. IMHO it would be desirable to drop that as well and perhaps they could be read from the workspace recommendations if they are not already.

That's something we should talk about broadly. The goal here was to add the default extensions needed to get up and running. While in this case it's just eslint, for python and others there's much more to it.  Consistency is important IMHO.

> To sum up, I think we should aim for "one-liners" containers.

I think there is a balancing act here between "fast", "usable", and ensuring people what to do next.  Generally, I find that people don't read documentation until they hit a problem.  By providing a template, we help people getting off the ground.

FWIW, the Salesforce container that was contributed also contains a largely blank Dockerfile to make it easy for people to customize. They're also using an image. 

Rather than focusing on one liners, I think a better approach would actually be to look into adopting the VSO kitchen sink image and letting people "open folder in container" without picking a template at all to "try" things out. In this mode, we can add some smarts.  Once they are ready to optimize their experience, these templates help them go farther.

@chrmarti - Your thoughts here?

**Comment by Chuxel** [comment]
@alexdima Oh, one more thing.

>Why is the above needed? Why is our built-in terminal not picking /bin/bash by default? IMHO we should fix things inside VS Code such that the terminal picks the best available shell, and not the most common one. Perhaps an issue already exists that we could prioritize?

On Linux, I am using `zsh` while others use `fish` and I have my VS Code  user settings set to these shells.  The problem is that these shells are not in all container images, so when you try to open a terminal, it fails saying the specified shell is not available.  This sets a machine level setting.  We could make it "null" as well just to be sure the setting doesn't apply.

This was added as a result of customer feedback and confusion on this exact topic.

**Comment by chrmarti** [comment]
I think it's important to keep the target audience for the definitions in mind. Most developers (except those doing deployments themselves) will have little experience with Docker and will be overwhelmed by too much as well as too little information on "where to go from here". I expect finding the right balance to be tricky.

I suggest we keep the Dockerfiles, they nicely demonstrate the setup most users will need if they go beyond kicking the tire. It's great that these could be simplified now that we patch the UID on Linux (for single containers). Maybe we can expand on that by avoiding `"terminal.integrated.shell.linux"` in the settings when we add `"terminal.integrated.shell.linux": null` to the machine settings automatically?

On the long apt-get command: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/ suggests a simplified version:
```Dockerfile
RUN apt-get update && apt-get install -y \
    aufs-tools \
    automake \
    build-essential \
 && rm -rf /var/lib/apt/lists/*
```
Would that be sufficient? (Maybe adding `--no-install-recommends`.)

**Comment by Chuxel** [comment]
> I expect finding the right balance to be tricky.

Yeah, agreed. The more I think about this, the more I think we may need something in the middle here.  We could have:

Kitchen Sink => "language / platform pointer" => Full devcontainer.json

Kitchen sink is tracked as #155 and we could clearly do some automation when this is selected.  We could then have something in the middle that's a straight pointer to one of these pre-defined definitions with a name that's basically the language and / or platform.  This would enable the "one liner" idea that Alex is suggesting for scenarios where that makes the most sense while still keeping the full power available.  We could then have some smarts in "reopen folder in container" that gives you the option to auto-select.  This is sort of buildpack-esque and is somewhat similar to the VS concept of a "workload".  That also means it's quite compatible with VSO's goals as well.

>On the long apt-get command: https://docs.docker.com/develop/develop-images/dockerfile_best-practices/ suggests a simplified version:

You get warnings in red in the build logs without setting the env, but we could append 2>&1 to avoid that problem so the warning goes to stdin instead of stderr.

> Maybe we can expand on that by avoiding "terminal.integrated.shell.linux" in the settings when we add `"terminal.integrated.shell.linux": null` to the machine settings automatically?

What I'm not sure about is what the least confusing behavior would be. Right now, it's pretty clear that the value is being overridden. So if you go into user settings and change it and it has no effect, you can see what is going on from devcontainer.json.  If the default is that the user setting isn't applied, that may not be clear.  I guess one question is whether the settings UI should default the scope to Remote when you are connected to a remote endpoint instead of User. That would fix that problem.

**Comment by chrmarti** [comment]
Do all packages trigger stderr output without these env variables set? It might be only those asking for user input. We could just remove them from the commented out examples for clarity and leave them in where we actually run them. (Is the bold red for stderr in the log overdoing it?)

"Machine" settings behave that way (or at least very similar): `"Configuration that can be configured only in the user settings when the extension is running locally, or only in the remote settings when the extension is running remotely."`

**Comment by alexdima** [comment]
Another possibility for avoiding:
```
	// Set *default* container specific settings.json values on container create.
	"settings": { 
		"terminal.integrated.shell.linux": "/bin/bash"
	},
```
in `devcontainer.json`

Would be if the base image, `mcr.microsoft.com/vscode/devcontainers/typescript-node:12`, would contain a file `settings.json` at the right location (we can ask @sandy081 and @chrmarti what that is) and the machine settings.json is predefined in the base image. Then, the effect of that is there, but it is not repeated in all the workspaces that want to use this image.

**Comment by Chuxel** [comment]
> Do all packages trigger stderr output without these env variables set? It might be only those asking for user input. We could just remove them from the commented out examples for clarity and leave them in where we actually run them. (Is the bold red for stderr in the log overdoing it?)

Yes, its an innocuous warning telling you that it is flipping to noninteractive, but it shows up in red.  Dirk actually raised an issue on it way back pre-launch which is when I adapted to it.  [See here](https://code.visualstudio.com/docs/remote/containers-advanced#_debconf-delaying-package-configuration-since-aptutils-is-not-installed) - I even documented it.

> Would be if the base image, mcr.microsoft.com/vscode/devcontainers/typescript-node:12, would contain a file settings.json at the right location (we can ask @sandy081 and @chrmarti what that is) and the machine settings.json is predefined in the base image. Then, the effect of that is there, but it is not repeated in all the workspaces that want to use this image.

That was what I was doing originally prior to the addition of the settings property actually.  Personally I don't see the harm in having it visible, since it also tells people they can change it to somethign else if they want.  That was really the problem with this approach - the setting was applied, but nobody was aware of it.

That gets to the same point earlier on the default - the question is really what the most sensible default should be.

The more we talk about this, the more I really think we need some sort of templating / project concept beyond image so you can just refer to an already defined devcontainer json.  Over time we could also let people define their own sets of templates for reuse.

**Comment by Chuxel** [comment]
This topic has broken out into several others and the info here is a bit stale.  Closing for now in favor of others more specific issues.

**Comment by leohxj** [comment]
how vsc-dev-containers implement `updateRemoteUserUID`?

In `docker run` docs, I see `user` args, only change `username:gid`, not specify `uid`
